### PR TITLE
Fix the ensemble_size == 0 error in automl.py

### PR DESCRIPTION
**Pull request by bkpcoding**
Fixing the ```ensemble_size == 0``` error in the ```fit_ensemble``` and ```show_models``` functions by raising a ValueError in the former and issuing a warning and returning an empty dictionary in the latter. Related to issue #1365

### CODE REVIEW

**Comment by bkpcoding** [comment]
Hey @eddiebergman can you check the PR?

File: autosklearn/automl.py [hunk 1]
```diff
@@ -1906,6 +1907,9 @@ def show_models(self) -> Dict[int, Any]:
         """
 
         ensemble_dict = {}
+        if self._ensemble_size == 0:
+            self._logger.warning('No models in the ensemble. Kindly check the ensemble size.')
```
Comment (eddiebergman) [inline autosklearn/automl.py:1911]: As this is a front facing API, this warning should be emitted to the user and not just logged silently, in fact there's not much point in logging it silently as it wouldn't help with debugging much.

```python
import warnings
warnings.warn(msg)
```
Also, I don't think `_ensemble_size` is the best proxy for checking this. A use could pass `_ensemble_size = 10` but auto-sklearn failed to find any models, resulting in `self._ensemble_size = 10` but having no `models_` property.

Take a look at [`_load_models()`](https://github.com/automl/auto-sklearn/blob/master/autosklearn/automl.py#L1510) for the logic of `self.ensemble_` and `self.models_`.

While thinking about it, there's probably three separate conditions to check for:
* Has autosklearn been fitted? If not raise a RuntimeError
* Has autosklearn been given a paramter of `ensemble_size=0`? If so, issue a warning and give an empty dict.
* Has autosklearn been given a parameter `ensemble_size > 0` but there is no ensemble to load from? If so, issue a warning and give an empty dict.
* Has 
I think checking `self.ensemble_` for `None` should be good enough

I would change it to
```
# At top
import warnings

if self._ensemble_size == 0
```

File: autosklearn/automl.py [hunk 2]
```diff
@@ -1455,8 +1455,9 @@ def fit_ensemble(self, y, task=None, precision=32,
                      dataset_name=None, ensemble_nbest=None,
                      ensemble_size=None):
         # AutoSklearn does not handle sparse y for now
+        if ensemble_size == 0:
+            raise ValueError("ensemble_size must be greater than 0 for fit_ensemble")
```
Comment (eddiebergman) [inline autosklearn/automl.py:1459]: Very small detail but match the check with the error message's statement :)
A little safer to have `> 0`

**Review by eddiebergman — DISMISSED** [review]
Seems mostly good :) Just a bit more explicit checking with `show_models()`.

We would also add some brief tests to make sure these are raised.
This constitutes 4 tests, one for each kind of error/warning. These would all go in `test_automl.py`

* Test that `fit_ensemble(..., ensemble_size=0)` gives an error. You don't need to worry about fitting before hand as the check is performed before anything explicit is done.
* Check for the fitted error in `show_models()`, should be straight forward enough.
* Check for empty dict when `show_models()` is called with `ensemble_size = 0` passed in construction of `AutoMLClassifier`.
* The last test, checking for no models existing return an empty dict, We could construct an `AutoMLClassifier` with `ensemble_size > 0`, set whatever flags are needed to indicate it has been fitted so it passed the other check and then test for the correct kind of error being raise.

For reference for tests if you're not familiar, you'll need `pytest.raises(<ErrorType>, match=<msg>)` for checking for errors being emitted.

**Comment by bkpcoding** [comment]
Thanks for the comments, will try to write tests and implement the suggestions.

**Comment by eddiebergman** [comment]
Hi @bkpcoding,

Do you plan to continue this? If not it's okay but I would rather close this if it's going to go stale.

Best,
Eddie

**Comment by bkpcoding** [comment]
Hey @eddiebergman,
I am extremely sorry, I have my exams going on but I did implement some of the changes you mentioned and will try to push them ASAP.

**Comment by eddiebergman** [comment]
No rush!! Please focus on your exams, this can wait. Just wanted to make sure there was still some plan to do this :)

**Comment by bkpcoding** [comment]
Hey, @eddiebergman I have made the changes you had requested and added two tests for them, however, it was not clear to me how to implement the points you suggested below, I tried a few things but was not able to successfully run them. 

- Has autosklearn been fitted? If not raise a RuntimeError
- The last test, checking for no models existing return an empty dict, We could construct an AutoMLClassifier with ensemble_size > 0, set whatever flags are needed to indicate it has been fitted so it passed the other check and then test for the correct kind of error being raise.
If you can help me with these. But if the changes are minor and you feel that the changes of the pr are not very important for now and might take your significant time, you can close the pr, I shall not mind.

File: autosklearn/automl.py [hunk 3]
```diff
@@ -968,8 +972,12 @@ def fit(
             self._logger.info("Finished loading models...")
 
         self._fit_cleanup()
+        self.fitted = True
 
         return self
+    
+    def __sklearn_is_fitted__(self)->bool:
```
Comment (eddiebergman) [inline autosklearn/automl.py:979]: You'll get an issue about the formatting here
```suggestion
    def __sklearn_is_fitted__(self) -> bool:
```

**Comment by eddiebergman** [comment]
Seems good to me, I'll resolve the conflicts and run the workflow files :) If there's any issues, you have to git pull and fix them but otherwise I'm happy with it!

@mfeurer, take a look if you like.

**Comment by eddiebergman** [comment]
I've done the merge, future contributions with formatting and checking should be easier with some new [tooling](https://github.com/automl/auto-sklearn/pull/1358) :) If you want to run the pre-commit checks locally (the ones that normally fail -_-), you can do:
```
pip install pre-commit
pre-commit run --all-files
```

I'll check back in later today and see how the tests do

**Comment by eddiebergman** [comment]
Sorry, rerunning tests, seems like one of them had pip installation issue but it seems unrelated to this (timeouts when trying to find a requirement). I reran them just to be sure

**Comment by eddiebergman** [comment]
Hey @bkpcoding,

Thanks very much for contributing! I've merged it so congrats :)

If you have any comments on how the contribution process was for you, we'd be happy to hear any thoughts you might have to help us improve for the future!
* How was interacting with our codebase/tests/GitHub actions?
* Did you use the Contribution guide and is there anything you'd change with it?
* How was the response/feedback during the contribution?
* Any thoughts on how we could improve for future contributions?
* Would you be interested in contributing again in any capacity?

You can leave it here as a comment or else you can find my email on my GitHub profile if you'd prefer to do so privately!

Thanks again!

Best,
Eddie

**Comment by bkpcoding** [comment]
Hey @eddiebergman thanks for merging the PR and for all the trouble you took for review and suggestions. I think the entire codebase was very well written and after reading the paper and going through the documentation I was pretty much able to grasp the implementation of the library.
The contribution guide was extremely useful though, I was able to set up the development environment without any hassle.
And any problems that I faced were due to my lack of complete understanding of git and git commands, however, a quick google was able to sort them out. 
And I would love to contribute to the library as much as I can. And sorry for the late replies as I had exams, but as they have ended I will be regularly checking the issues.

**Comment by eddiebergman** [comment]
Great to hear :) The code base can always be improved but step by step!

Git is always one of those things that takes time, I still don't know much of it's functionality but always plenty more to learn. If there's anything you found helpful to know about git, feel free to throw it into the Contribution guide.

If you like, I can tag you on issues that I don't think require going too deep or require extensive changes, otherwise feel free to poke around whenever you feel like it!

Best of luck with the exams!

Best,
Eddie

**Comment by bkpcoding** [comment]
Surely you can tag me on the issues you think would be relevant to me, I would be happy to look at them

**Pull request merged**